### PR TITLE
added distanceSquared function 

### PR DESCRIPTION
**Pull request by eraffxi**
to allow for lighter distance checks against squared values rather than finding the square root in distance()

### CODE REVIEW

**Comment by teschnei** [comment]
where would you use this?

File: src/common/utils.cpp [hunk 1]
```diff
@@ -79,6 +79,27 @@ float distance(const position_t& A, const position_t& B)
 *																		*
 ************************************************************************/
 
+float distanceSquared(const position_t& A, const position_t& B)
+{
+	float one = 0, two = 0, three = 0, four = 0;
```
Comment (xdemolish) [inline src/common/utils.cpp]: distance_x, distance_y, distance_z

**Comment by xdemolish** [comment]
@teschnei 
```
<warp> so my friend took a look at the mobroam stuff and said the distance checking is not as optimized as it could be
<warp> he says instead of doing a square root in the distance check, it is way less work for a cpu to square a number than find the square root
<warp> he suggested doing a distanceSquared() function that just returns an absolute value and checking it against a squared version of the original check
<warp> so distanceSquared() <= 100 instead of distance() <= 10
```

**Comment by eraffxi** [comment]
i honestly am not 100% sure the best way to use it and am testing it with the distance checks for mob skills by using
          
  if (currentDistance <= pow(PMobSkill->getDistance(), 2))
  where currentDistance = distanceSquared(PMob->loc.p, PActionTarget->loc.p);

but i'm not sure if that's actually faster or slower than just doing PMobSkill->getDistance() * PMobSkill->getDistance()

max said he's in favor of it so i figured i'd offer it to be used by someone who might know faster than me how best to use it

**Comment by teschnei** [comment]
he's probably right, but why does it matter?

**Comment by eraffxi** [comment]
i'm using it because the analysis i took showed 50-60% of all cpu time was spent on mobroam and hozu suggested it might be due to checking distance for ranged attacks, so i'm testing that.  since max said 'no reason not to use it really' i pushed this for him

**Comment by teschnei** [comment]
if that's actually what the problem is, then we can consider using this (or something similar to this that's easier to use)
but you need to make sure that's actually the problem, because I can't realistically see that being the case (why would a mob be checking for ranged attacks while roaming?)

**Comment by Deadwing888** [comment]
Mobs also don't roam if no one is in the zone. That's not retail but I assume it was a choice made to conserve processing power.

For example if you are the only person in the zone and you pull an NM 500' away from its spawn and log out, it'll still be right next to you when you log back in, no matter how long you wait.

**Comment by Hozu** [comment]
The problem I mentioned is that every tick it's checking attempting to actually use the ability (provided it's not on cooldown) - if it doesn't have a target, there's no distance to check.

**Comment by bendangelo** [comment]
Sure the function is faster and thanks for the work. If we want to improve the performence of mob roaming we have to do profiling. I've never done that before in C++

**Comment by xdemolish** [comment]
@hozu thought i fixed that in https://github.com/DarkstarProject/darkstar/issues/3120 ?

**Comment by Hozu** [comment]
Well they lose their target with the fix, but it doesn't stop them from keep trying to shoot. They just don't shoot with no target.

**Comment by teschnei** [comment]
TrySpecialSkill fails nearly immediately if there's no target

 I'm just going to close this for now unless it can actually be proven (ie using visual studio's profiler) that this SPECIFICALLY is a bottleneck

**Pull request closed without merging**